pyteal/ast/scratch.py

```diff
@@ -17,26 +17,36 @@ class ScratchSlot:
     # Slot ids under 256 are manually reserved slots
     nextSlotId = NUM_SLOTS
 
-    def __init__(self, requestedSlotId: int = None):
+    def __init__(self, requestedSlotId: int = None, *, slotIdFromStack: bool = False):
         """Initializes a scratch slot with a particular id
 
         Args:
             requestedSlotId (optional): A scratch slot id that the compiler must store the value.
             This id may be a Python int in the range [0-256).
         """
-        if requestedSlotId is None:
-            self.id = ScratchSlot.nextSlotId
-            ScratchSlot.nextSlotId += 1
-            self.isReservedSlot = False
+        assert (
+            requestedSlotId is None or not slotIdFromStack
+        ), "cannot specify requestedSlotId when slotIdFromStack"
+
+        self.slotIdFromStack = slotIdFromStack
+
+        if self.slotIdFromStack:
+            self.id = -1
+            self.isReservedSlot = True
```

Comment:
nit: in this condition branch, `requestedSlotId` is unchecked, while I would love to enforce it to be `None` to make things clearer.

Reply:
I'll add an extra assertion. Even though in theory, you can never get to this branch when `requestedSlotId != None` given the assertion on line 28 above, explicit is better than implicit.

Reply:
oh no, I overlooked things...

Reply:
ok, will not add the extra check